### PR TITLE
PLT-5012 Combine updateLastViewedAt, setLastViewedAt and setActiveChannel into a single API

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
* Deprecate updateLastViewedAt, setLastViewedAt and setActiveChannel endpoints
* Add a new `/channels/view` endpoint to consolidate all the functionality of the above endpoints into

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5012
https://mattermost.atlassian.net/browse/PLT-5013

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (mattermost/mattermost-api-reference#57)
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

File: model/client.go [hunk 1]
```diff
@@ -48,6 +48,13 @@ type Result struct {
 	Data      interface{}
 }
 
+type ResponseMetadata struct {
```
Comment (enahum) [inline model/client.go]: good one!

**Comment by jwilander** [comment]
Oops yeah, I forgot to split it up. Will do next time

File: api/channel.go [hunk 2]
```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	w.Write([]byte(channels.ToJson()))
 }
+
+func viewChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+	view := model.ChannelViewFromJson(r.Body)
+
+	go func() {
```
Comment (crspeller) [inline api/channel.go]: If this ends up being slow this `go` will cause problems. Please remove.

File: api/channel.go [hunk 3]
```diff
@@ -965,7 +866,27 @@ func getChannel(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		status.ActiveChannel = channelId
+		if !status.Manual {
+			status.Status = model.STATUS_ONLINE
+		}
+		status.LastActivityAt = model.GetMillis()
+	}
+
+	AddStatusCache(status)
+
+	if result := <-Srv.Store.Status().SaveOrUpdate(status); result.Err != nil {
```
Comment (crspeller) [inline api/channel.go]: Do we really have to send this to the DB? 

How about only sending to the DB for manual statuses? You can update the cache when the server starts and every time anyone changes their manual status. Automatic status is probably the common case.

File: api/channel.go [hunk 4]
```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
...
+				return
+			}
+		}
+
+		if len(view.PrevChannelId) > 0 {
+			Srv.Store.Channel().UpdateLastViewedAt(view.PrevChannelId, c.Session.UserId)
+
+			// Only clear push notifications if a channel switch occured
+			if *utils.Cfg.EmailSettings.SendPushNotifications && !c.Session.IsMobileApp() {
+				go func() {
```
Comment (crspeller) [inline api/channel.go]: I think we are moving away from having the notifications stuff spawn more go routines.

File: api/channel.go [hunk 5]
```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		}
+
+		teamPref := model.Preference{
+			UserId:   c.Session.UserId,
+			Category: model.PREFERENCE_CATEGORY_LAST,
+			Name:     model.PREFERENCE_NAME_LAST_TEAM,
+			Value:    c.TeamId,
+		}
+
+		Srv.Store.Preference().Save(&model.Preferences{teamPref, chanPref})
```
Comment (crspeller) [inline api/channel.go]: What do we use this for? There are so many writes in here.

Comment (enahum) [inline api/channel.go]: This are use to switch to the last team/channel for the team sidebar

Comment (crspeller) [inline api/channel.go]: So you just added this with the team sidebar stuff?

File: api/channel.go (hunk 3 above)
Comment (jwilander) [inline api/channel.go]: This isn't new code I just moved it here. That said, I think the DB write can be completely removed in this instance now that the changes in statuses are broadcast across HA. If the status is automatic or manual shouldn't matter since that should have been taken care of by the normal method of statuses being set.

File: api/channel.go (hunk 5 above)
Comment (jwilander) [inline api/channel.go]: We discussed and `BrowserStore` will be used. I will remove the preference writes from this PR.

File: api/channel.go [hunk 6]
```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
...
+			}
+		} else {
+			if result := <-Srv.Store.Channel().SetLastViewedAt(view.ChannelId, c.Session.UserId, view.Time); result.Err != nil {
+				c.Err = result.Err
+				return
+			}
+		}
+
+		if len(view.PrevChannelId) > 0 {
+			Srv.Store.Channel().UpdateLastViewedAt(view.PrevChannelId, c.Session.UserId)
```
Comment (coreyhulen) [inline api/channel.go]: We should maybe combine the 2 UpdateLastViewedAt Calls into 1 sql statement?  This can be a separate PR.  Someting like `WHERE Id=ID1 OR Id=ID2`

File: model/client.go [hunk 7]
```diff
@@ -1341,6 +1349,24 @@ func (c *Client) UpdateLastViewedAt(channelId string, active bool) (*Result, *Ap
 	}
 }
 
+// ViewChannel performs all the actions related to viewing a channel. This includes marking
+// the channel and the previous one as read, marking the channel as being actively viewed.
+// ChannelId is required but may be blank to indicate no channel is being viewed.
+// PrevChannelId is optional, populate to indicate a channel switch occurred. Optionally
+// provide a non-zero Time, in Unix milliseconds, to manually set the viewing time.
+func (c *Client) ViewChannel(params ChannelView) (bool, *ResponseMetadata) {
```
Comment (coreyhulen) [inline model/client.go]: I think we wanted to return the time? and use this to fix some other race issue?  (I think)

File: api/channel.go [hunk 8]
```diff
@@ -1293,3 +1214,66 @@ func autocompleteChannels(c *Context, w http.ResponseWriter, r *http.Request) {
...
+
+	if len(view.ChannelId) > 0 {
+
+		if view.Time == 0 {
+			if result := <-Srv.Store.Channel().UpdateLastViewedAt(view.ChannelId, c.Session.UserId); result.Err != nil {
+				c.Err = result.Err
+				return
+			}
+		} else {
+			if result := <-Srv.Store.Channel().SetLastViewedAt(view.ChannelId, c.Session.UserId, view.Time); result.Err != nil {
```
Comment (coreyhulen) [inline api/channel.go]: I don't think we should trust the time of the user?  It needs to be server time and passed back to the user.

File: model/client.go (hunk 7 above)
Comment (jwilander) [inline model/client.go]: That would require another DB read to actually pull the time (since we use `LastPostAt` on the channel if a time isn't provided) but if needed I can add it

File: api/channel.go (hunk 6 above)
Comment (jwilander) [inline api/channel.go]: Will be addressed with https://mattermost.atlassian.net/browse/PLT-5073

File: api/channel.go (hunk 2 above)
Comment (jwilander) [inline api/channel.go]: Will be addressed with https://mattermost.atlassian.net/browse/PLT-5073

File: api/channel.go (hunk 4 above)
Comment (jwilander) [inline api/channel.go]: Will be addressed with https://mattermost.atlassian.net/browse/PLT-5073

File: api/channel.go (hunk 3 above)
Comment (jwilander) [inline api/channel.go]: Will be addressed with https://mattermost.atlassian.net/browse/PLT-5073

File: api/channel.go (hunk 8 above)
Comment (jwilander) [inline api/channel.go]: We were already trusting the user time in the original `set_last_viewed_at` API, I just moved the functionality here. What's wrong with trusting the user time? It's only setting their own unreads right

File: model/client.go (hunk 7 above)
Comment (jwilander) [inline model/client.go]: @enahum and I already messaged briefly about this

File: api/channel.go (hunk 8 above)
Comment (coreyhulen) [inline api/channel.go]: Is this the unread messages issue?

**Comment by coreyhulen** [comment]
Also more a general comment, but can you get your mos recent channel with `SELECT * FROM Channels WHERE Channel.Id = ChannelMember.ChannelId AND ChannelMember.User = 'MyID' ORDER BY ChannelMembers.LastUpdateAt DESC LIMIT 1`

Just a thought.  Could remove the one in the status table.  Can also probably remove the one in preference too. 0/5

**Comment by coreyhulen** [comment]
I'm ok if my comments are address in another PR... +1 for now.

File: api/channel.go (hunk 8 above)
Comment (jwilander) [inline api/channel.go]: Nope, we only use user time for that one feature where you could mark a channel as unread by alt clicking on a post
